# Exploring Engagement on Hacker News: Ask HN vs Show HN

Hacker News is a popular platform started by Y Combinator, where users submit posts that can receive votes and comments, similar to Reddit. It's especially influential in the tech and startup world, and posts that reach the top can attract hundreds of thousands of views.

For this project, we're working with a dataset of about 20,000 posts, which has been reduced from an original 300,000 by filtering out posts without comments and then randomly sampling the rest. Each post in the dataset includes information like the title, author, number of points (upvotes minus downvotes), number of comments, and the time it was created.

We'll specifically focus on two types of posts: "Ask HN," where users ask the community a question, and "Show HN," where users share a project or product. Our goal is to find out:

- Do "Ask HN" or "Show HN" posts tend to get more comments on average?
- Does the time a post is created impact how many comments it receives?

By exploring these questions, we'll gain insights into how engagement works on Hacker News.

In [1]:
from csv import reader

opened_file = open('hacker_news.csv') #open the file
read_file = reader(opened_file) #read in the csv format
hn = list(read_file) #create a list from the csv file

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[0] #list containing just the header
hn = hn[1:] #removing the header from the HackerNews Dataset

print(headers)
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Filtering Data

We are only interested in post titles with "Ask HN" or "Show HN" so we will filter the data and create new lists of list containing the data for those titles. 

In [83]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('The length of ask_posts list is {}'.format(len(ask_posts)))
print('The length of the show_posts list is {}'.format(len(show_posts)))
print('The length of the other_posts list is {}'.format(len(other_posts)))
print('\n')
print('First 5 rows of ask_posts:')
print(ask_posts[:5])
print('\n')
print('First 5 rows of show_posts:')
print(show_posts[:5])

The length of ask_posts list is 1744
The length of the show_posts list is 1162
The length of the other_posts list is 17194


First 5 rows of ask_posts:
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


First 5 rows of show_posts:
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', '

### Analyzing Comments Per Posts

In [49]:
total_ask_comments = 0

#average number of comments in ask_posts
for row in ask_posts:
    comment = row[4]
    comment = int(comment)
    total_ask_comments += comment

avg_ask_comments = total_ask_comments / len(ask_posts)
print('The average number of comments in ask posts is {}.'.format(avg_ask_comments))


#average nunmber of comments in show posts
total_show_comments = 0

for row in show_posts:
    s_comment = row[4]
    s_comment = int(s_comment)
    total_show_comments += s_comment

avg_show_comments = total_show_comments / len(show_posts)
print('The average number of comments in show posts is {}.'.format(avg_show_comments))

The average number of comments in ask posts is 14.038417431192661.
The average number of comments in show posts is 10.31669535283993.


Ask posts seem to receive on average more comments than show posts so we'll focus on those for now. We will see if asks posts that are created at certain times are more likely to receive comments. To do this, we'll create a frequency table to calculate the number of ask posts created in a each hour of the day and the corresponding number of comments received. We will then calculate the average number of comments ask posts received by each hour. 

### Ask Posts Comments Frequency Table

In [61]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    comment = int(row[4])
    result_list.append([created_at, comment])

counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

for row in result_list:
    date = row[0]
    hour = dt.datetime.strptime(date, date_format).strftime('%H')
    comment  = row[1]
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
        
print("Counts By Hour:")
print(counts_by_hour)
print('\n')
print("Comments By Hour:")
print(comments_by_hour)

Counts By Hour:
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Comments By Hour:
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


Using the count_by_hour and comments_by_hour dictionaries, we will now calculate the average number of comments for posts created during each hour of the day

In [73]:
avg_by_hour = []

for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

Let's clean up this list so it's easier to see the hours with the highest values (average number of comments).

In [79]:
#swapping the columns of avg_by_hour
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

#sorting the swapped list in descending order
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [82]:
print("Top 5 Hours For Ask Posts Comments:")
for row in sorted_swap[0:5]:
    hour = row[1]
    hour = dt.datetime.strptime(hour, '%H').strftime('%H:00')
    print('{}: {:.2f} averages comments per post'.format(hour, row[0]))

Top 5 Hours For Ask Posts Comments:
15:00: 38.59 averages comments per post
02:00: 23.81 averages comments per post
20:00: 21.52 averages comments per post
16:00: 16.80 averages comments per post
21:00: 16.01 averages comments per post


The hour with the highest average number of comments per post is 15:00 (3:00 PM EST), with an average of 38.59 comments per post. This represents roughly a 60% increase in comments compared to the second highest hour. According to the dataset documentation, the time is in Eastern Time (EST).

### Show Posts Comments Frequency Table

In [85]:
import datetime as dt

result_list = []

for row in show_posts:
    created_at = row[6]
    comment = int(row[4])
    result_list.append([created_at, comment])

counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

for row in result_list:
    date = row[0]
    hour = dt.datetime.strptime(date, date_format).strftime('%H')
    comment  = row[1]
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
        
print("Counts By Hour:")
print(counts_by_hour)
print('\n')
print("Comments By Hour:")
print(comments_by_hour)

Counts By Hour:
{'14': 86, '22': 46, '18': 61, '07': 26, '20': 60, '05': 19, '16': 93, '19': 55, '15': 78, '03': 27, '17': 93, '06': 16, '02': 30, '13': 99, '08': 34, '21': 47, '04': 26, '11': 44, '12': 61, '23': 36, '09': 30, '01': 28, '10': 36, '00': 31}


Comments By Hour:
{'14': 1156, '22': 570, '18': 962, '07': 299, '20': 612, '05': 58, '16': 1084, '19': 539, '15': 632, '03': 287, '17': 911, '06': 142, '02': 127, '13': 946, '08': 165, '21': 272, '04': 247, '11': 491, '12': 720, '23': 447, '09': 291, '01': 246, '10': 297, '00': 487}


In [86]:
avg_by_hour = []

for hour in counts_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

#swapping the columns of avg_by_hour
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

#sorting the swapped list in descending order
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
sorted_swap

print("Top 5 Hours For Show Posts Comments:")
for row in sorted_swap[0:5]:
    hour = row[1]
    hour = dt.datetime.strptime(hour, '%H').strftime('%H:00')
    print('{}: {:.2f} averages comments per post'.format(hour, row[0]))

Top 5 Hours For Show Posts Comments:
18:00: 15.77 averages comments per post
00:00: 15.71 averages comments per post
14:00: 13.44 averages comments per post
23:00: 12.42 averages comments per post
22:00: 12.39 averages comments per post


### Analyzing Points Per Posts

In [84]:
total_ask_points = 0

#average number of points in ask_posts
for row in ask_posts:
    points = row[3]
    points = int(points)
    total_ask_points += points

avg_ask_points = total_ask_points / len(ask_posts)
print('The average number of points in ask posts is {}.'.format(avg_ask_points))


#average nunmber of points in show posts
total_show_points = 0

for row in show_posts:
    s_points = row[3]
    s_points = int(s_points)
    total_show_points += s_points

avg_show_points = total_show_points / len(show_posts)
print('The average number of points in show posts is {}.'.format(avg_show_points))

The average number of points in ask posts is 15.061926605504587.
The average number of points in show posts is 27.555077452667813.


Unlike with comments, show posts receive more points on average than ask posts. 

### Show Posts Points Frequency Table

In [90]:
import datetime as dt

result_list = []

for row in show_posts:
    created_at = row[6]
    points = int(row[3])
    result_list.append([created_at, points])

counts_by_hour = {}
points_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

for row in result_list:
    date = row[0]
    hour = dt.datetime.strptime(date, date_format).strftime('%H')
    points  = row[1]
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        points_by_hour[hour] = points
    else:
        counts_by_hour[hour] += 1
        points_by_hour[hour] += points
        
print("Counts By Hour:")
print(counts_by_hour)
print('\n')
print("Points By Hour:")
print(points_by_hour)

Counts By Hour:
{'14': 86, '22': 46, '18': 61, '07': 26, '20': 60, '05': 19, '16': 93, '19': 55, '15': 78, '03': 27, '17': 93, '06': 16, '02': 30, '13': 99, '08': 34, '21': 47, '04': 26, '11': 44, '12': 61, '23': 36, '09': 30, '01': 28, '10': 36, '00': 31}


Points By Hour:
{'14': 2187, '22': 1856, '18': 2215, '07': 494, '20': 1819, '05': 104, '16': 2634, '19': 1702, '15': 2228, '03': 679, '17': 2521, '06': 375, '02': 340, '13': 2438, '08': 519, '21': 866, '04': 386, '11': 1480, '12': 2543, '23': 1526, '09': 553, '01': 700, '10': 681, '00': 1173}


In [94]:
avg_by_hour = []

for hour in counts_by_hour:
    avg_by_hour.append([hour, points_by_hour[hour]/counts_by_hour[hour]])


#swapping the columns of avg_by_hour
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

#sorting the swapped list in descending order
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print("Top 5 Hours For Show Posts Comments:")
for row in sorted_swap[0:5]:
    hour = row[1]
    hour = dt.datetime.strptime(hour, '%H').strftime('%H:00')
    print('{}: {:.2f} averages points per post'.format(hour, row[0]))

Top 5 Hours For Show Posts Comments:
23:00: 42.39 averages points per post
12:00: 41.69 averages points per post
22:00: 40.35 averages points per post
00:00: 37.84 averages points per post
18:00: 36.31 averages points per post


The hour with the most Show Posts comments is 23:00 (11:00 PM EST) with an average of 42.39 points per post. 